lib/filewatch/fingerprinter.rb

```diff
@@ -0,0 +1,130 @@
+require 'filewatch/boot_setup' unless defined?(FileWatch)
+
+module FileWatch
+  class Fingerprinter
+    attr_reader :data, :fnv, :fingerprint, :offset, :size
+
+    def initialize(path, offset)
+      @path = path
+      @offset = offset
+      @size = FP_BYTE_SIZE
+    end
+
+    def match_any?(keys)
+      if @fnv.nil?
+        # STDERR.puts "Fingerprinter: fnv nil for path - #{path}"
+        return nil
+      end
+      if @data.nil?
+        # STDERR.puts "Fingerprinter: data nil for path - #{path}"
+        return nil
+      end
+      keys.each do |k|
+        fp, off, len = *k
+        # STDERR.puts "Fingerprinter: any_match? - #{fp}:#{off}:#{len}"
+        if off != @offset
+          # STDERR.puts "Fingerprinter: any_match? - wrong offset, mine is: #{@offset}"
+          next
+        end
+        possible = @fnv.fnv1a64(len)
+        if fp == possible
+          @fingerprint, @size = fp, len
+          return k
+        end
+      end
+      # STDERR.puts "Fingerprinter: no match - #{keys.inspect}"
+      nil
+    end
+
+    def add_size(size)
+      if @fnv
+        fnv_fingerprint(size)
+      end
+      self
+    end
+
+    def add_data(data)
+      @data = data
+      set_fnv
+      self
+    end
+
+    def read_path
+      begin
+        file = FileOpener.open(@path)
+        set_fingerprint(file)
+      ensure
+        file.close
+      end
+      self
+    end
+
+    def read_file(file)
+      set_fingerprint(file)
+      self
+    end
```

Comment:
This method has side effect, since it will move the file pointer after reading the file, it is the intended behavior?
I have looked where it is used but this that case it was harmless since we are closing the FD after generating the fingerprint.

If its the intended behavior maybe add a comment here and add a bang `!` at the end? WDYT?